File: dotnet/Services/Migration/Coverage.cs
using System.Text.Json.Nodes;

namespace PanCopilot.Services.Migration;

// Port of migration/coverage.py. Values: true=implemented, false=planned, "partial".
public static class Coverage
{
    public static readonly Dictionary<string, Dictionary<string, object>> Matrix = new()
    {
        ["cisco"] = new() { ["objects"] = true, ["groups"] = true, ["services"] = true, ["security"] = true, ["nat"] = true, ["interfaces"] = true, ["routes"] = true, ["vpn"] = true },
        ["checkpoint"] = new() { ["objects"] = true, ["groups"] = "partial", ["services"] = true, ["security"] = true, ["nat"] = "partial", ["interfaces"] = "partial", ["routes"] = "partial", ["vpn"] = false },
        ["fortinet"] = new() { ["objects"] = true, ["groups"] = false, ["services"] = true, ["security"] = true, ["nat"] = false, ["interfaces"] = false, ["routes"] = false, ["vpn"] = false },
        ["juniper"] = new() { ["objects"] = true, ["groups"] = false, ["services"] = false, ["security"] = true, ["nat"] = false, ["interfaces"] = false, ["routes"] = false, ["vpn"] = false },
        ["palo"] = new() { ["objects"] = true, ["groups"] = "partial", ["services"] = "partial", ["security"] = true, ["nat"] = "partial", ["interfaces"] = false, ["routes"] = false, ["vpn"] = false },
    };

    private static readonly string[] Features =
        { "objects", "groups", "services", "security", "nat", "interfaces", "routes", "vpn" };

    public static JsonObject Snapshot()
    {
        var vendors = new JsonObject();
        foreach (var (vendor, feats) in Matrix)
        {
            var f = new JsonObject();
            foreach (var (k, v) in feats)
                f[k] = v switch { bool b => b, string s => s, _ => v.ToString() };
            vendors[vendor] = f;
        }
        var features = new JsonArray();
        foreach (var f in Features) features.Add(f);
        return new JsonObject { ["vendors"] = vendors, ["features"] = features };
    }
}

File: local/migration/coverage.py
"""Feature coverage matrix for Config Migration (Expedition-style)."""

from __future__ import annotations

# Values: True = implemented in engine, False = planned, "partial" = basic support
COVERAGE_MATRIX: dict[str, dict[str, bool | str]] = {
    "cisco": {
        "objects": True,
        "groups": True,
        "services": True,
        "security": True,
        "nat": True,
        "interfaces": True,
        "routes": True,
        "vpn": True,
    },
    "checkpoint": {
        "objects": True,
        "groups": "partial",
        "services": True,
        "security": True,
        "nat": "partial",
        "interfaces": "partial",
        "routes": "partial",
        "vpn": False,
    },
    "fortinet": {
        "objects": True,
        "groups": False,
        "services": True,
        "security": True,
        "nat": False,
        "interfaces": False,
        "routes": False,
        "vpn": False,
    },
    "juniper": {
        "objects": True,
        "groups": False,
        "services": False,
        "security": True,
        "nat": False,
        "interfaces": False,
        "routes": False,
        "vpn": False,
    },
    "palo": {
        "objects": True,
        "groups": "partial",
        "services": "partial",
        "security": True,
        "nat": "partial",
        "interfaces": False,
        "routes": False,
        "vpn": False,
    },
}

FEATURE_LABELS = [
    ("objects", "Address objects"),
    ("groups", "Object groups"),
    ("services", "Services"),
    ("security", "Security policy"),
    ("nat", "NAT"),
    ("interfaces", "L3 interfaces"),
    ("routes", "Static routes"),
    ("vpn", "VPN"),
]


def coverage_for_vendor(vendor: str) -> dict[str, bool | str]:
    return COVERAGE_MATRIX.get(vendor, {})


def coverage_snapshot() -> dict:
    return {"vendors": COVERAGE_MATRIX, "features": [f[0] for f in FEATURE_LABELS]}